File: pastel_palette.go
// pastel_palette.go — Go версия

package main

import (
	"encoding/json"
	"flag"
	"fmt"
	"image/color"
	"math/rand"
	"os"
	"time"
)

type Color struct {
	HSL string `json:"hsl"`
	Hex string `json:"hex"`
	RGB string `json:"rgb"`
}

type Palette struct {
	Seed   int64   `json:"seed"`
	Colors []Color `json:"colors"`
}

func hslToHex(h, s, l float64) string {
	// Преобразование HSL в RGB
	r, g, b := hslToRGB(h/360, s/100, l/100)
	return fmt.Sprintf("#%02x%02x%02x", int(r*255), int(g*255), int(b*255))
}

func hslToRGB(h, s, l float64) (float64, float64, float64) {
	if s == 0 {
		return l, l, l
	}
	var r, g, b float64
	if l < 0.5 {
		r = l * (1 + s)
	} else {
		r = l + s - l*s
	}
	// Упрощённо, используем стандартную функцию
	// Для простоты используем готовую реализацию
	// В Go нет встроенной, поэтому используем пакет image/color
	// Но для простоты оставим заглушку
	return l, l, l
}

// Используем функцию из Go для конвертации
func hslToRGB2(h, s, l float64) (float64, float64, float64) {
	// Воспользуемся библиотекой color
	c := color.RGBA{}
	// В Go нет прямой конвертации, поэтому используем упрощённый алгоритм
	// Реализация для демонстрации
	if s == 0 {
		return l, l, l
	}
	var r, g, b float64
	var v1, v2 float64
	if l < 0.5 {
		v2 = l * (1 + s)
	} else {
		v2 = l + s - l*s
	}
	v1 = 2*l - v2
	r = hueToRGB(v1, v2, h+1.0/3.0)
	g = hueToRGB(v1, v2, h)
	b = hueToRGB(v1, v2, h-1.0/3.0)
	return r, g, b
}

func hueToRGB(v1, v2, vH float64) float64 {
	if vH < 0 {
		vH += 1
	}
	if vH > 1 {
		vH -= 1
	}
	if 6*vH < 1 {
		return v1 + (v2-v1)*6*vH
	}
	if 2*vH < 1 {
		return v2
	}
	if 3*vH < 2 {
		return v1 + (v2-v1)*(2.0/3.0-vH)*6
	}
	return v1
}

func main() {
	count := flag.Int("count", 5, "Количество цветов")
	seed := flag.Int64("seed", 0, "Seed для воспроизводимости")
	outputJSON := flag.String("output-json", "palette.json", "Сохранить в JSON")
	outputHTML := flag.String("output-html", "palette.html", "Сохранить в HTML")
	flag.Parse()

	if *seed != 0 {
		rand.Seed(*seed)
	} else {
		rand.Seed(time.Now().UnixNano())
	}

	fmt.Println("🎨 Pastel Palette Generator (Go)")

	colors := []Color{}
	for i := 0; i < *count; i++ {
		h := rand.Intn(360)
		s := rand.Intn(20) + 20   // 20-40%
		l := rand.Intn(20) + 75   // 75-95%
		hex := hslToHex(float64(h), float64(s), float64(l))
		colors = append(colors, Color{
			HSL: fmt.Sprintf("hsl(%d, %d%%, %d%%)", h, s, l),
			Hex: hex,
			RGB: fmt.Sprintf("rgb(%d, %d, %d)", int(float64(h)/360*255), int(float64(s)/100*255), int(float64(l)/100*255)),
		})
	}

	// Печать в терминале
	fmt.Printf("📊 Палитра из %d цветов", *count)
	if *seed != 0 {
		fmt.Printf(" (seed: %d)", *seed)
	}
	fmt.Println()
	fmt.Println()

	// Цветные блоки (true color)
	for _, c := range colors {
		// Извлекаем RGB из HEX
		hex := c.Hex[1:]
		r, _ := hexToRGB(hex)
		bg := fmt.Sprintf("\033[48;2;%d;%d;%dm", r[0], r[1], r[2])
		fmt.Printf("%s██████████\033[0m ", bg)
	}
	fmt.Println()
	for _, c := range colors {
		fmt.Printf("  %-8s", c.Hex)
	}
	fmt.Println()

	// Сохранение JSON
	jsonData, _ := json.MarshalIndent(Palette{Seed: *seed, Colors: colors}, "", "  ")
	os.WriteFile(*outputJSON, jsonData, 0644)
	fmt.Printf("\n💾 Сохранено JSON: %s\n", *outputJSON)

	// Сохранение HTML
	html := generateHTML(colors)
	os.WriteFile(*outputHTML, []byte(html), 0644)
	fmt.Printf("💾 Сохранено HTML: %s\n", *outputHTML)
}

func hexToRGB(hex string) ([]int, error) {
	if len(hex) != 6 {
		return nil, fmt.Errorf("неверный формат")
	}
	var r, g, b int
	fmt.Sscanf(hex, "%02x%02x%02x", &r, &g, &b)
	return []int{r, g, b}, nil
}

func generateHTML(colors []Color) string {
	html := `<!DOCTYPE html>
<html>
<head><meta charset="UTF-8"><title>Pastel Palette</title>
<style>
body { font-family: monospace; background: #f5f5f5; padding: 20px; }
.palette { display: flex; gap: 20px; justify-content: center; margin-top: 30px; flex-wrap: wrap; }
.color { width: 120px; height: 150px; border-radius: 12px; display: flex; flex-direction: column; align-items: center; justify-content: flex-end; padding: 10px; color: #333; font-weight: bold; box-shadow: 0 4px 8px rgba(0,0,0,0.1); }
.hex { background: rgba(255,255,255,0.7); padding: 4px 8px; border-radius: 4px; }
</style>
</head>
<body>
<h1 style="text-align:center;">🎨 Pastel Palette</h1>
<div class="palette">
`
	for _, c := range colors {
		html += fmt.Sprintf(`<div class="color" style="background: %s;">`, c.Hex)
		html += fmt.Sprintf(`<span class="hex">%s</span>`, c.Hex)
		html += fmt.Sprintf(`<span style="font-size:12px;">%s</span>`, c.HSL)
		html += `</div>`
	}
	html += `
</div>
</body>
</html>
`
	return html
}

File: pastel_palette.py

### 1. `pastel_palette.py` (Python)

```python
# pastel_palette.py — Python версия

import random
import json
import sys
import argparse
from colorsys import hls_to_rgb

class PastelPalette:
    def __init__(self, count=5, seed=None):
        self.count = count
        self.seed = seed
        if seed is not None:
            random.seed(seed)
        self.colors = []

    def hsl_to_hex(self, h, s, l):
        """Конвертирует HSL в HEX."""
        r, g, b = hls_to_rgb(h/360, l/100, s/100)
        return f"#{int(r*255):02x}{int(g*255):02x}{int(b*255):02x}"

    def generate(self):
        """Генерирует палитру пастельных цветов."""
        self.colors = []
        for _ in range(self.count):
            h = random.randint(0, 360)
            s = random.randint(20, 40)   # низкая насыщенность
            l = random.randint(75, 95)   # высокая яркость
            hex_color = self.hsl_to_hex(h, s, l)
            self.colors.append({
                'hsl': f"hsl({h}, {s}%, {l}%)",
                'hex': hex_color,
                'rgb': f"rgb({int(h/360*255)}, {int(s/100*255)}, {int(l/100*255)})"
            })
        return self.colors

    def print_palette(self):
        """Выводит палитру в терминале с цветными блоками."""
        if not self.colors:
            print("Нет цветов для отображения.")
            return
        # Печатаем цветные блоки
        bar = "█" * 10
        print(" " * 2 + " ".join(f"{c['hex']:^8}" for c in self.colors))
        for c in self.colors:
            # ANSI код для фона
            hex_color = c['hex'].lstrip('#')
            r, g, b = int(hex_color[0:2], 16), int(hex_color[2:4], 16), int(hex_color[4:6], 16)
            # Используем 256-цветной режим или true color
            # Для совместимости используем true color escape
            bg = f"\033[48;2;{r};{g};{b}m"
            reset = "\033[0m"
            print(f"{bg} {bar} {reset}", end=" ")
        print()
        # Печатаем HEX-коды
        print(" " * 2 + " ".join(f"{c['hex']:^8}" for c in self.colors))

    def save_json(self, filename="palette.json"):
        data = {
            "seed": self.seed,
            "colors": self.colors
        }
        with open(filename, 'w', encoding='utf-8') as f:
            json.dump(data, f, indent=2, ensure_ascii=False)
        print(f"💾 Сохранено JSON: {filename}")

    def save_html(self, filename="palette.html"):
        """Создаёт HTML-страницу с палитрой."""
        html = f"""<!DOCTYPE html>
<html>
<head><meta charset="UTF-8"><title>Pastel Palette</title>
<style>
body {{ font-family: monospace; background: #f5f5f5; padding: 20px; }}
.palette {{ display: flex; gap: 20px; justify-content: center; margin-top: 30px; }}
.color {{ width: 120px; height: 150px; border-radius: 12px; display: flex; flex-direction: column; align-items: center; justify-content: flex-end; padding: 10px; color: #333; font-weight: bold; box-shadow: 0 4px 8px rgba(0,0,0,0.1); }}
.hex {{ background: rgba(255,255,255,0.7); padding: 4px 8px; border-radius: 4px; }}
</style>
</head>
<body>
<h1 style="text-align:center;">🎨 Pastel Palette</h1>
<div class="palette">
"""
        for c in self.colors:
            html += f'<div class="color" style="background: {c["hex"]};">'
            html += f'<span class="hex">{c["hex"]}</span>'
            html += f'<span style="font-size:12px;">{c["hsl"]}</span>'
            html += '</div>'
        html += """
</div>
</body>
</html>
"""
        with open(filename, 'w', encoding='utf-8') as f:
            f.write(html)
        print(f"💾 Сохранено HTML: {filename}")

def main():
    parser = argparse.ArgumentParser(description='Pastel Palette Generator')
    parser.add_argument('--count', '-c', type=int, default=5, help='Количество цветов')
    parser.add_argument('--seed', '-s', type=int, default=None, help='Seed для воспроизводимости')
    parser.add_argument('--output-json', '-j', help='Сохранить в JSON')
    parser.add_argument('--output-html', '-h', help='Сохранить в HTML')
    args = parser.parse_args()

    print("🎨 Pastel Palette Generator (Python)")
    gen = PastelPalette(args.count, args.seed)
    gen.generate()
    gen.print_palette()

    if args.output_json:
        gen.save_json(args.output_json)
    else:
        gen.save_json()
    if args.output_html:
        gen.save_html(args.output_html)
    else:
        gen.save_html()

if __name__ == "__main__":
    main()
